File: 79-word-search/word-search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        n=len(board)
        m=len(board[0])
        vis=[[False]*m for _ in range(n)]
        mark=[]
        for i in range(n):
            for j in range(m):
                if board[i][j]==word[0]:
                    mark.append((i,j))
        if len(mark)==0:
            return False
        def backtrack(i,j,idx):
            if idx==len(word):
                return True
            direction=[(1,0),(0,1),(-1,0),(0,-1)]
            for x,y in direction:
                nx,ny=i+x,j+y
                if 0<= nx <n and 0 <= ny < m and not vis[nx][ny] and board[nx][ny]==word[idx]:
                    vis[nx][ny]=True
                    if backtrack(nx, ny, idx + 1):
                        return True

                    vis[nx][ny]=False
            return False
                    
        for i,j in mark:
            vis=[[False]*m for _ in range(n)]
            vis[i][j]=True
            if backtrack(i,j,1):
                return True
            
        return False
        

File: 79-word-search/word-search.cpp
class Solution {
public:
    bool dfs(vector<vector<char>>& board, string word,int i,int j,int k){
        if(k>=word.size()) return true;
        if(i<0 || j<0 || i>=board.size() || j>=board[0].size() || board[i][j]!=word[k]) return false;
        char c=board[i][j];
        board[i][j]='#';
        bool ans=dfs(board,word,i+1,j,k+1) || dfs(board,word,i-1,j,k+1) || dfs(board,word,i,j+1,k+1) || dfs(board,word,i,j-1,k+1);
        board[i][j]=c;
        return ans;
    }
    bool exist(vector<vector<char>>& board, string word) {
        int m=board.size(),n=board[0].size();
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(board[i][j]==word[0]){
                    if(dfs(board,word,i,j,0)) return true;
                }
            }
        }
        return false;
    }
};